### ISSUE TITLE
Documentation/guidance for composing multiple middlewares together

### ISSUE THREAD

**Issue by bilalq** [middyjs/middy#1036]
**Is your feature request related to a problem? Please describe.**

I'd love guidance on how to compose multiple middlewares together into a single wrapper middleware.

I find that there are multiple middlewares used across almost all of our lambdas. One common set of groupings is the one from powertools:

```typescript
export const handler = middy<APIGatewayProxyEvent, APIGatewayProxyResult>()
  .use(
    captureLambdaHandler(tracer, {
      captureResponse: true,
    })
  )
  .use(
    injectLambdaContext(logger, {
      clearState: true,
      logEvent: false,
    })
  )
  .use(
    logMetrics(metrics, {
      captureColdStartMetric: true,
    })
  )
  .handler(async (event, context) => {
    // ...
  })
```

I'd much rather write it as:

```typescript
export const handler = middy<APIGatewayProxyEvent, APIGatewayProxyResult>()
  .use(
    powertools({
      tracerOptions: {
        captureResponse: true,
      },
      loggerOptions: {
        clearState: true,
        logEvent: false,
      },
      metricsOptions: {
        captureColdStartMetric: true,
      },
    })
  )
  .handler(async (event, context) => {
    // ...
  })
```

To that end, I've created a middleware that composes them like so:

```typescript
import { Logger, injectLambdaContext } from '@aws-lambda-powertools/logger'
import { Metrics, logMetrics } from '@aws-lambda-powertools/metrics'
import { Tracer, captureLambdaHandler } from '@aws-lambda-powertools/tracer'
import middy from '@middy/core'

export const logger = new Logger()
export const metrics = new Metrics()
export const tracer = new Tracer()

/**
 * Options for the composed powertools middleware. Composes over base middleware
 * options but makes them non-nullable so that consumers are forced to be aware
 * of all powertools that will be used.
 */
export interface PowerToolsOptions {
  tracer: NonNullable<Parameters<typeof captureLambdaHandler>[1]>
  logger: NonNullable<Parameters<typeof injectLambdaContext>[1]>
  metrics: NonNullable<Parameters<typeof logMetrics>[1]>
}

/**
 * Unified middleware that composes over tracer, logger, and metrics Lambda
 * Powertools middlewares.
 */
export const powertools = (options: PowerToolsOptions): middy.MiddlewareObj => {
  const powertoolsMiddlewares = [
    captureLambdaHandler(tracer, options.tracer),
    injectLambdaContext(logger, options.logger),
    logMetrics(metrics, options.metrics),
  ]

  const composeMiddlewarePhase = (key: keyof middy.MiddlewareObj) => {
    return async (request: middy.Request) => {
      for (const middleware of powertoolsMiddlewares) {
        await middleware[key]?.(request)
      }
    }
  }

  return {
    before: composeMiddlewarePhase('before'),
    after: composeMiddlewarePhase('after'),
    onError: composeMiddlewarePhase('onError'),
  }
}
```

The question I have is whether or not my serial execution of middlewares in the order of application is correct. Does that mirror what middy does if one just chains `.use` calls? If the goal is to mimic the chained `.use` behavior, do I need `try/catch` handling to make sure some of these middlewares run even if earlier ones in the chain error out? Or is it correct as-is?

Beyond this, I know it'd be a little more complicated if the middlewares were modifying the request/response/context/error types in the chain, so documentation for that would also be nice.

**Describe the solution you'd like**

Documentation or guidance on how to do this.

**Describe alternatives you've considered**

Tried implementation described above.

**Additional context**

N/A

**Comment by willfarrell** [comment]
You're not the first to ask about this.  There are two options:

1. Have middleware function that includes middy (recommended)

```
// shared file
export const middify = (options) => {
  return middy()
  .use(
    captureLambdaHandler(tracer, { ...options.tracer,
      captureResponse: true,
    })
  )
  .use(
    injectLambdaContext(logger, {...options.logger,
      clearState: true,
      logEvent: false,
    })
  )
  .use(
    logMetrics(metrics, {...options.metrics,
      captureColdStartMetric: true,
    })
  )

}

// lambda file
import {middify} from 'above'
export const handler = middify(...).use(...).handler(...)
```

2. merge into a single middleware

This is what you have above, it should work as expect. It matches what middy does under the the hood expect for plugin hooks (which few use) and the ability to short circuit the request early by returning a value (few middleware need).

This is what middy runs on an array of middlewares
```
const runMiddlewares = async (request, middlewares, plugin) => {
  for (const nextMiddleware of middlewares) {
    plugin.beforeMiddleware?.(nextMiddleware.name)
    const res = await nextMiddleware(request)
    plugin.afterMiddleware?.(nextMiddleware.name)
    // short circuit chaining and respond early
    if (typeof res !== 'undefined') {
      request.response = res
      return
    }
  }
}
```

**Comment by bilalq** [comment]
Thanks, that's helpful. I'd love to use option 1, but it seems like that limits me to just a single grouping. I may have `powertools` as one composed middleware and another composed one that's common across all lambdas within a specific service.

Would you be open to an API for first-class composing? Maybe an easy way to do this would be to have `.use` overloaded to accept an array of middlewares? Or have a function called `composeMiddlewares` that does the merging as part of middy's core. Admittedly, the typing of this gets a little complicated.

**Comment by Crafoord** [comment]
> `.use` already supports accepting an array of middlewares :)

I've been struggling to get this concept to work in a nice way when using typescript. When I use an array the types are instantly gone. It works great if i have a "static" set of middlewares (no array) but after using middy for several years that is rarely the case, often you have different needs for different endpoints (some are authed, some use validator etc etc).

```ts
// above: creating arrays with a bunch of middlewares based on options...
if (options.apiLambda) {
    return middy().use([...baseMiddlewares, ...apiMiddlewares]);
  }
return middy().use(baseMiddlewares);
```

Have done several iterations but it never quite gets right, the array approach would have been amazing since it's easy to combine a specific array based on some options for the middify function. Maybe this is more a limitation in typescript...

Usually end up having to create my own generic types to figure out what will be the "output" after all middlewares have run which is a shame since it should be there already.

**Comment by willfarrell** [comment]
Improvement to types are always welcome.
